packages/SwingSet/src/kernel/state/vatKeeper.js

```diff
@@ -148,33 +228,70 @@ export function makeVatKeeper(
     return harden(options);
   }
 
-  function updateReapInterval(reapInterval) {
-    reapInterval === 'never' ||
-      isNat(reapInterval) ||
-      Fail`bad reapInterval ${reapInterval}`;
-    kvStore.set(`${vatID}.reapInterval`, `${reapInterval}`);
-    if (reapInterval === 'never') {
-      kvStore.set(`${vatID}.reapCountdown`, 'never');
-    }
-  }
+  // This is named "addDirt" because it should increment all dirt
+  // counters (both for reap/BOYD and for heap snapshotting). We don't
+  // have `heapSnapshotDirt` yet, but when we do, it should get
+  // incremented here.
 
-  function countdownToReap() {
-    const rawCount = getRequired(`${vatID}.reapCountdown`);
-    if (rawCount === 'never') {
-      return false;
-    } else {
-      const count = Number.parseInt(rawCount, 10);
-      if (count === 1) {
-        kvStore.set(
-          `${vatID}.reapCountdown`,
-          getRequired(`${vatID}.reapInterval`),
-        );
-        return true;
-      } else {
-        kvStore.set(`${vatID}.reapCountdown`, `${count - 1}`);
-        return false;
+  /**
+   * Add some "dirt" to the vat, possibly triggering a reap/BOYD.
+   *
+   * @param {Dirt} moreDirt
+   */
+  function addDirt(moreDirt) {
+    // this references our cached 'reapDirt' and 'reapDirtThreshold'
+    assert.typeof(moreDirt, 'object');
+    let reap = false;
+    for (const key of Object.keys(moreDirt)) {
+      const threshold = reapDirtThreshold[key];
+      // Don't accumulate dirt if it can't eventually trigger a
+      // BOYD. This is mainly to keep comms from counting upwards
+      // forever. TODO revisit this when we add heapSnapshotDirt,
+      // maybe check both thresholds and accumulate the dirt if either
+      // one is non-'never'.
+      if (threshold && threshold !== 'never') {
+        const oldDirt = reapDirt[key] || 0;
+        // The 'moreDirt' value might be Number or BigInt (eg
+        // .computrons). We coerce to Number so we can JSON-stringify.
```

Comment:
I thought we converted the `computrons` to Number before putting the value in the dirt record?

Reply:
There's a duplicate coercion, yeah, I figured that was harmless.

This one was annoying, I first tried to make `gcKrefs` be a BigInt, to match how we handle computrons, but then I can't use simple JSON to store `reapDirt` or `reapDirtThreshold`, and holding it in RAM as a BigInt but converting back/forth to Number in the save/load functions was messy and distracting. Also, my choice to use `key is missing` as "no dirt of that category" made things slightly harder (the `|| 0` is to ignore `gcKrefs: undefined` -like cases).

And I didn't want to go back and change `computrons` to be Number everything, that was too much churn. I decided that I'd treat computrons as potentially large, and gcKrefs as always pretty small, and use that to justify leaving computrons as BigInts in the existing code. But then I ignore that argument in `addDirt` and do not tolerate large computron sizes. Sigh.

Reply:
Oh yeah `Number` is totally fine here for dirt computrons. Even if we go over the MAX_SAFE_INTEGER value, the loss of precision is deterministic and doesn't impact the threshold logic.